sourmash/index.py

```diff
@@ -0,0 +1,103 @@
+"An Abstract Base Class for collections of signatures."
+
+from abc import ABCMeta, abstractmethod
+from collections import namedtuple
+
+# @CTB copied out of search.py to deal with import order issues, #willfix
+SearchResult = namedtuple('SearchResult',
+                          'similarity, match_sig, md5, filename, name')
+
+
+# compatible with Python 2 *and* 3:
+ABC = ABCMeta("ABC", (object,), {"__slots__": ()})
+
+
+class Index(ABC):
+    @abstractmethod
+    def find(self, search_fn, *args, **kwargs):
+        """ """
+
+    @abstractmethod
+    def search(self, signature, *args, **kwargs):
+        """ """
+
+    @abstractmethod
+    def gather(self, signature, *args, **kwargs):
+        """ """
+
+    @abstractmethod
+    def insert(self, node):
+        """ """
+
+    @abstractmethod
+    def save(self, path, storage=None, sparseness=0.0, structure_only=False):
+        """ """
+
+    @classmethod
+    @abstractmethod
+    def load(cls, location, leaf_loader=None, storage=None, print_version_warning=True):
+        """ """
+
+
+class LinearIndex(Index):
+    def __init__(self):
+        self.signatures = set()
+
+    def insert(self, node):
+        self.signatures.add(node)
+
+    def find(self, search_fn, *args, **kwargs):
+        matches = []
+
+        for node in self.signatures:
+            if search_fn(node, *args):
+                matches.append(node)
+        return matches
+
+    def search(self, query, *args, **kwargs):
+        """@@
+
+        Note, the "best only" hint is ignored by LinearIndex.
+        """
+
+        # check arguments
+        if 'threshold' not in kwargs:
+            raise TypeError("'search' requires 'threshold'")
+        threshold = kwargs['threshold']
+
+        do_containment = kwargs.get('do_containment', False)
+        ignore_abundance = kwargs.get('ignore_abundance', False)
```

Comment:
It's fine to define args here:
```python
    def search(self, query, *args,
                     threshold=None,
                     do_containment=False,
                     ignore_abundance=False, 
                     **kwargs):
         """@@
 
         Note, the "best only" hint is ignored by LinearIndex.
         """
 
         if threshold is None:
             raise TypeError("'search' requires 'threshold'")
```

The issue is it defeats a bit the purpose of having the `Index` abstraction: should we move these args into the `Index` instead (especially since they have defaults), and then it's up for any class implementing `Index` to use them or not?

(I would like to see the `search` and `gather` commands defined in terms of `Index` instead of SBT/LCA/LinearIndex specifics)

Reply:
yes, I agree we should do that. wending my way through the Python syntax is all :)